clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+            sslEngine.setUseClientMode(mode == Mode.CLIENT);
+            appBuffer = ByteBuffer.allocate(sslEngine.getSession().getApplicationBufferSize());
+            netBuffer = ByteBuffer.allocate(sslEngine.getSession().getPacketBufferSize());
+        }
+
+        void beginHandshake() throws SSLException {
+            sslEngine.beginHandshake();
+        }
+
+        void handshake(SSLConfigValidatorEngine peerEngine) throws SSLException {
+            SSLEngineResult.HandshakeStatus handshakeStatus = sslEngine.getHandshakeStatus();
+            while (true) {
+                switch (handshakeStatus) {
+                    case NEED_WRAP:
+                        handshakeResult = sslEngine.wrap(EMPTY_BUF, netBuffer);
+                        switch (handshakeResult.getStatus()) {
+                            case OK: break;
+                            case BUFFER_OVERFLOW:
+                                netBuffer.compact();
+                                Utils.ensureCapacity(netBuffer, sslEngine.getSession().getPacketBufferSize());
+                                netBuffer.flip();
+                                break;
+                            case BUFFER_UNDERFLOW: throw new SSLException("Unexpected status: CLOSED");
+                            case CLOSED: throw new SSLException("Unexpected status: CLOSED");
+                        }
+                        return;
+                    case NEED_UNWRAP:
+                        if (peerEngine.netBuffer.position() == 0)
+                            return;
+                        peerEngine.netBuffer.flip();
+                        handshakeResult = sslEngine.unwrap(peerEngine.netBuffer, appBuffer);
+                        peerEngine.netBuffer.compact();
+                        handshakeStatus = handshakeResult.getHandshakeStatus();
+                        switch (handshakeResult.getStatus()) {
+                            case OK: break;
+                            case BUFFER_OVERFLOW:
+                                Utils.ensureCapacity(appBuffer, sslEngine.getSession().getApplicationBufferSize());
+                                break;
+                            case BUFFER_UNDERFLOW:
+                                Utils.ensureCapacity(netBuffer, sslEngine.getSession().getPacketBufferSize());
+                                break;
+                            case CLOSED: throw new SSLException("Unexpected status: CLOSED");
+                        }
+                        break;
+                    case NEED_TASK:
+                        sslEngine.getDelegatedTask().run();
+                        handshakeStatus = sslEngine.getHandshakeStatus();
+                        break;
+                    case FINISHED:
+                        return;
+                    case NOT_HANDSHAKING:
+                        if (handshakeResult.getHandshakeStatus() != SSLEngineResult.HandshakeStatus.FINISHED)
```

Comment:
Can you explain why we need this check? Aren't we already checking the result of `getHandshakeStatus()`?

Reply:
I am not sure we hit this case, but I wanted to be cautious. It is because both handshake result and sslEngine store handshake status and they may not be the same.